repo/repo.go

```diff
@@ -1,282 +1,331 @@
+
+// ServeHTTP performs strict matching on the repo part of the URL path and
+// dispatches the request to the appropriate handler.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	urlPath := r.URL.Path
+	if urlPath == "/" {
+		// TODO: add HEAD and GET
+		switch r.Method {
+		case "POST":
+			h.createRepo(w, r)
+		default:
+			httpMethodNotAllowed(w, []string{"POST"})
+		}
+		return
+	} else if urlPath == "/config" {
+		switch r.Method {
+		case "HEAD":
+			h.checkConfig(w, r)
+		case "GET":
+			h.getConfig(w, r)
+		case "POST":
+			h.saveConfig(w, r)
+		case "DELETE":
+			h.deleteConfig(w, r)
+		default:
+			httpMethodNotAllowed(w, []string{"HEAD", "GET", "POST", "DELETE"})
+		}
+		return
+	} else if objectType, objectID := h.getObject(urlPath); objectType != "" {
+		if objectID == "" {
+			// TODO: add HEAD
+			switch r.Method {
+			case "GET":
+				h.listBlobs(w, r)
+			default:
+				httpMethodNotAllowed(w, []string{"GET"})
+			}
+			return
+		} else {
+			switch r.Method {
+			case "HEAD":
+				h.checkBlob(w, r)
+			case "GET":
+				h.getBlob(w, r)
+			case "POST":
+				h.saveBlob(w, r)
+			case "DELETE":
+				h.deleteBlob(w, r)
+			default:
+				httpMethodNotAllowed(w, []string{"HEAD", "GET", "POST", "DELETE"})
+			}
+			return
+		}
 	}
-	return join(s.Path, s.getRepo(r), fileType)
+	httpDefaultError(w, http.StatusNotFound)
 }
 
-// getFilePath returns the path for a file in the repo.
-func (s *Server) getFilePath(r *http.Request, fileType, name string) (string, error) {
-	if !s.isValidType(fileType) {
-		return "", errors.New("invalid file type")
+// getObject parses the URL path and returns the objectType and objectID,
+// if any. The objectID is optional.
+func (h *Handler) getObject(urlPath string) (objectType, objectID string) {
+	if m := BlobPathRE.FindStringSubmatch(urlPath); len(m) > 0 {
+		if len(m) == 2 || m[2] == "" {
+			return m[1], ""
+		}
+		return m[1], m[2]
+	} else {
+		return "", ""
 	}
+}
 
-	if s.isHashed(fileType) {
-		if len(name) < 2 {
-			return "", errors.New("file name is too short")
-		}
+// getSubPath returns the path for a file or subdir in the root of the repo.
+func (h *Handler) getSubPath(name string) string {
+	return filepath.Join(h.path, name)
+}
 
-		return join(s.Path, s.getRepo(r), fileType, name[:2], name)
+// getObjectPath returns the path for an object file in the repo.
+// The passed in objectType and objectID must be valid due to earlier validation
+func (h *Handler) getObjectPath(objectType, objectID string) string {
+	// If we hit an error, this is a programming error, because all of these
+	// must have been validated before. We still check them here as a safeguard.
+	if objectType == "" || objectID == "" {
+		panic("invalid objectType or objectID")
+	}
+	if isHashed(objectType) {
+		if len(objectID) < 2 {
+			// Should never happen, because BlobPathRE checked this
+			panic("getObjectPath: objectID shorter than 2 chars")
```

Comment:
I'd add the `objectID` to the string passed to `panic()` so that we get a bit of information should this ever happen :)

Reply:
Right, I will add it.